### DISCUSSION TITLE
The idea behind django-cypress

### DISCUSSION THREAD

**Discussion by Stavrospanakakis** [HelloWorld-PC/django-cypress#2] [General]
### Summary
The purpose of this issue is to discuss the concept and goals behind the `django-cypress` project, which aims to bring the power of end-to-end testing and behavior-driven development to [Django](https://www.djangoproject.com/) projects using the [Cypress](https://www.cypress.io/) framework.

### Background
The [Cypress](https://www.cypress.io/) framework has gained significant popularity among web developers, especially in the [Laravel](https://laravel.com/) community, with the [`laracasts/cypress`](https://github.com/laracasts/cypress) package providing seamless integration and additional utilities for [Laravel](https://laravel.com/) projects. However, there is currently no similar project available for [Django](https://www.djangoproject.com/), which led to the inception of `django-cypress`.

### Example Ideas
1. [`cy.refreshDatabase()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L214-L216): This method cleans the database and runs migrations, ensuring a clean and consistent state before running tests. It helps in preventing interference from previous test runs and facilitates reliable and isolated test execution.
2. [`cy.create()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L143-L204): This method enables the creation of new records for Django models within the testing environment. It simplifies the process of setting up test data, allowing developers to focus on writing test cases rather than manual data preparation.
3. [`cy.python()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L274-L301): This method provides the ability to execute Python code within the Cypress test environment. It enables developers to perform complex backend operations, interact with Django APIs, or access Django-specific functionality seamlessly during test execution.
4. [`cy.login()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L10-L34): This method facilitates user authentication and login within the Django project being tested. It streamlines the process of simulating user actions and interactions, enabling comprehensive end-to-end testing scenarios that involve authenticated user flows.

You can find a full list of the [laracasts/cypress](https://github.com/laracasts/cypress) in this [file](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js).

### Discussion
We welcome your thoughts, feedback, and suggestions on the proposed ideas and the overall direction of `django-cypress`. Please feel free to contribute your insights, share alternative approaches, or raise any concerns you may have regarding this initiative.

Let's work together to make `django-cypress` a valuable addition to the Django ecosystem!

**Comment by Stavrospanakakis** [comment]
I believe that the v1.0.0 should contain the following commands:
- `cy.manage()`: This will run the `python manage.py <args>` command. It is the equivalent of [`cy.artisan()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L245-L272) command.
- `cy.refreshDatabase()`: This will delete all the database tables and then it will run the `python manage.py migrate` command. It is the equivalent of [`cy.refreshDatabase()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L214-L216C4) command.
- `cy.login()`: This command will authenticate a user inside the tests. It is the equivalent of [`cy.login()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L10-L34) command.
- `cy.logout()`: This command will force the user to logout inside the tests. It is the equivalent of [`cy.logout()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L68-L82) command.

Reading the code from the Laravel Integration, I observed that we also need to create a [`cy.csrftoken()`](https://github.com/laracasts/cypress/blob/86b523e313c37e6b156131b63a90fc3a1197c5c4/src/stubs/support/laravel-commands.js#L89-L97) to make the above commands to work.

@fakiolinho @stsivoula 
Thoughts?

**Reply by fakiolinho** [reply]
@Stavrospanakakis great analysis!!

IMHO, In order to be able to use `django-cypress` and do some baby steps we need to offer to the library's users the ability to create models.

## A basic scenario

That said, I think we need to cover first a simple scenario like the one described below so we are able to test an actual web app with `cypress` for non-auth-protected views:

- run migrations and clear the DB data with a `django-cypress` command so we start from a clean slate when needed
- run seeders or even better create some very specific model instances like 10 blog posts instead of seeding fully the DB with a `django-cypress` command
- go to `/posts` view with a `cypress` command
- check that the `posts` lists actually render with a `cypress` command
- click on a delete option with a `cypress` command
- check how the UI behaves with a `cypress` command etc

## Clearing the DB

Maybe we can create 2 different commands to separate migrations and clearing the DB commands, so we avoid running the migrations again and again since mostly we'd like to clear the DB before a testcase.

## Next Steps

Once we put in place some very basic commands regarding migrations, seeders, and models creation, we can ship the very first release so we can actually use it across some existing Python Django web apps.

Right after, we can focus some more on the `auth` commands since the complexity there is higher.

These are the methods we probably need for auth flows:

- login method
- authUser method so we can access the logged-in user in our test cases
- logout method
- csrf method

What do you think about these 2 distinguished steps in our roadmap?

**Reply by Stavrospanakakis** [reply]
I like the idea of creating 2 separate commands for running migrations and clearing the database. 

By the way, Django does not provide a `seed` command. An engineer should create and run a custom command with the `python manage.py <subcommand>` command. For example, `python manage.py seed`. It is the equivalent of `php artisan <subcommand>`.

So I guest the proposal is to create the following commands for the v1.0.0, right?
- `cy.migrate()`: Run the migrations
- `cy.refreshDatabase()`: Clear the database
- `cy.manage()`: Run a `python manage.py` command

Thoughts? 
@fakiolinho @stsivoula
